include/ginkgo/core/base/array.hpp

```diff
@@ -627,6 +710,40 @@ class copy_back_deleter<Array<T>> {
 };
 
 
+/**
+ * @internal
+ *
+ * Casts away const-ness from a array.
+ *
+ * This preserves is_owning() and re-wraps deleters, and can be used to
+ * implement functionality for classes that only store non-const arrays.
+ *
+ * @param array  the array to be cast. It will be empty with default deleter
+ *               afterwards.
+ * @returns a non-const array wrapping the same pointer and deleter on the same
+ *          executor as `array`.
+ */
+template <typename ValueType>
+Array<ValueType> array_const_cast(Array<const ValueType>&& array)
+{
+    auto deleter = array.data_.get_deleter();
+    std::function<void(ValueType[])> new_deleter;
+    if (array.is_owning()) {
+        new_deleter = typename Array<ValueType>::default_deleter{array.exec_};
+    } else {
+        new_deleter = [deleter](auto ptr) { deleter(ptr); };
+    }
+    Array<ValueType> result{array.exec_, array.num_elems_,
+                            const_cast<ValueType*>(array.data_.release()),
+                            new_deleter};
+    array.num_elems_ = 0;
+    array.data_ = typename Array<const ValueType>::data_manager{
+        nullptr, typename Array<const ValueType>::default_deleter{array.exec_}};
```

Comment:
If a const array doesn't own the data, we could use the view deleter directly. This prevents compiler errors from trying to use `executor_deleter<const T>`.
```suggestion
        nullptr, typename Array<const ValueType>::const_view_deleter{}};
```